content/blog/prototyping/index.md

```diff
@@ -0,0 +1,726 @@
+If you play by the rules, it can be a bit cumbersome to pass around references and mutable references.
+The compiler is constantly reminding you of ownership and lifetimes and that can break your flow.
+
+```rust
+// First attempt with references - compiler error!
+struct Note<'a> {
+    title: &'a str,
+    content: &'a str,
+}
+
+fn create_note() -> Note<'_> {  // ❌ lifetime error
+    let title = String::from("Draft");
+    let content = String::from("My first note");
+    Note {
+        title: &title,
+        content: &content
+    }
+}
+```
+
+This code doesn't compile because the references are not valid outside of the function.
+
+```rust
+   Compiling playground v0.0.1 (/playground)
+error[E0106]: missing lifetime specifier
+ --> src/lib.rs:7:26
+  |
+7 | fn create_note() -> Note<'_> {  // ❌ lifetime error
+  |                          ^^ expected named lifetime parameter
+  |
+  = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
+help: consider using the `'static` lifetime, but this is uncommon unless you're returning a borrowed value from a `const` or a `static`, or if you will only have owned values
+  |
+```
+
+([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7274f20a06226316c93d9984f1d66b5f))
+
+A simple way around that is to avoid lifetimes altogether.
+They are not necessary in the beginning.
+Use owned types like `String` and `Vec` and you can always refactor later.
+Also, use `.clone()` wherever you need to get around it.
+
+```rust
+// Much simpler with owned types
+struct Note {
+    title: String,
+    content: String,
+}
+
+fn create_note() -> Note {  // ✓ just works
+    Note {
+        title: String::from("Draft"),
+        content: String::from("My first note")
+    }
+}
+```
+
+If you have a type that you need to move between threads (i.e. it needs to be `Send`), you can use `Arc<Mutex<T>>` to get around the borrow checker.
+If you're worried about performance, remember that other languages like Python or Java do this implicitly behind your back.
+
+```rust
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+let note = Arc::new(Mutex::new(Note {
+    title: String::from("Draft"),
+    content: String::from("My first note")
+}));
+
+let note_clone = Arc::clone(&note);
+thread::spawn(move || {
+    let mut note = note_clone.lock().unwrap();
+    note.content.push_str(" with additions");
+});
+```
+
+([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4b93a53ebc1d7ee6bc2b39c91543fba7))
+
+### Use A Flat Module Hierarchy
+
+Your `main.rs` is a great place for the entire prototype. 
+That's fine because you can always split it up later as soon as this becomes "uncomfortable."
+By then, you have a clear vision of what things should be called and how they should interact.
+
+I keep modules in the same file until they grow too large:
```

Comment:
By this point I didn't expect to use modules, but to have all functions and structs in the same scope. Maybe rephrase the header or previous paragraph? Seeing the `mod` keyword in the example surprised me a little.

Reply:
Agreed, if you have the capacity to think of modules , at that point just put them n a different file. Would refactor the whole point to say it's fine to just dump the whole code in a `main`

Reply:
Yeah, the headline needs to be changed. I will just say "flat hierarchy."
I still believe that there's a point to be made about the `mod` keyword being helpful for prototypes as well. I like to sketch out the module structure in `main`. It's great that Rust allows me to nest modules in a single file. It allows for quicker iteration cycles, because it's easier to move stuff around and at the same time I can build up a quick prototype of the hierarchy to see how it "feels." So I think I'll keep that, but mention it later in the section.

Reply:
Done